### ISSUE TITLE
Commands intermittently hang and are unresponsive

### ISSUE THREAD

**Issue by zx8** [fish-shell/fish-shell#5882]
Originally raised on Gitter: https://gitter.im/fish-shell/fish-shell?at=5ce26cefef853135c8efe521

Using https://github.com/fish-shell/fish-shell/commit/8e640cdcc549bd6a14a7a04f5354d75356303992 / macOS 10.14.5 / iTerm2 3.3.0beta8. 

Commands seem to intermittently hang and become unresponsive – that is, they don't respond to `CTRL+C` (SIGINT) or `CTRL+\` (SIGQUIT). I'm not sure how to debug this or provide more information, but am happy to so do if anyone has any suggestions.

Rolling back to https://github.com/fish-shell/fish-shell/commit/ff3fe961f4aa8d997ea2f65e309162161fc7df71 seems to fix the issue, so I suspect it's related to @ridiculousfish's recent commits.

**[edit]** Seems to bisect to https://github.com/fish-shell/fish-shell/commit/cd7e8f41033ea76c68b7f03cad466fd04ca9a03e.

**Comment by zx8** [comment]
Just ran into this again after executing `ls` but this time I managed to run `/usr/bin/sample` and grab a call graph for the hanging `fish` process if it helps...

```
Call graph:
    2671 Thread_402840   DispatchQueue_1: com.apple.main-thread  (serial)
      2671 start  (in libdyld.dylib) + 1  [0x7fff5ea613d5]
        2671 main  (in fish) + 5112  [0x101c1f004]
          2671 reader_read(int, io_chain_t const&)  (in fish) + 693  [0x101cb7c02]
            2671 reader_data_t::readline(int)  (in fish) + 500  [0x101cb71fa]
              2671 reader_data_t::read_normal_chars(readline_loop_state_t&)  (in fish) + 83  [0x101cb46d5]
                2671 input_readch(bool)  (in fish) + 530  [0x101c87e7f]
                  2671 input_common_readch()  (in fish) + 455  [0x101c8b287]
                    2671 __select  (in libsystem_kernel.dylib) + 10  [0x7fff5eb9d61a]
```

**Comment by ByScripts** [comment]
Seems I have the same problem since few days.

This can happens on any command, but really random.

For example, sometimes just a `cd my-dir` will hang. But for me `CTRL+C` will work, and I'll be in the `my-dir` directory, so the command is working.

Another problem appeared at the same time : Whenever I launch terminal, I'm in `/` directory instead of my home dir by default.

**Comment by faho** [comment]
@zx8: I can't reproduce. Can you try seeing if there's anything in your config that triggers this?

**Comment by zx8** [comment]
@faho My config is quite convoluted so I'm struggling to pinpoint exactly what's causing this, but I've confirmed it bisects to https://github.com/fish-shell/fish-shell/commit/cd7e8f41033ea76c68b7f03cad466fd04ca9a03e

**Comment by faho** [comment]
>grab a call graph for the hanging fish process if it helps...

That call graph just shows fish reading input. If it were stuck, it would be stuck in the same way that #5883 is - no binding, so it doesn't react to any input, even though it receives it. But I have no idea how that should happen?

#5883 is weirdness related to variable handlers, but even if they fire, they shouldn't be changing the bindings willy-nilly (#5883 is specifically about changing the bindings).

So I don't think the graph is relevant, and it's just coincidence. Most likely you hit the wrong thread.

>My config is quite convoluted so I'm struggling to pinpoint exactly what's causing this

I'd first look at the definition of `ls` - did you change it? Can you cause it with `command ls`? Then, I'd look at anything involving `source`, bindings or event handlers.

Alternatively, bisection is a useful technique - comment half of your config and try it. If it doesn't happen anymore, the trigger is in there.

**Comment by zx8** [comment]
@faho I have `ls` aliased to [`exa`](https://github.com/ogham/exa) but I'm not convinced the "hang" is anything to do with the actual command itself.

Upon further testing, it seems that sometimes this occurs before I've even hit <enter> to execute the command itself! I'm really stumped as to what it could be, but it certainly seems to have started occurring with commit https://github.com/fish-shell/fish-shell/commit/cd7e8f41033ea76c68b7f03cad466fd04ca9a03e as I can't reproduce this weird behaviour prior to that.

Anyhow,  I'll do as you say and try to bisect my configuration.

**Comment by faho** [comment]
>Upon further testing, it seems that sometimes this occurs before I've even hit <enter> to execute the command itself!

Here's a few possibilities:

- Suggestions trigger loading the completions (so we know which arguments can be files). Since it's an alias, this triggers loading the `exa` completions, not ls
- A binding that executes something that hangs.
- Either of those triggering an event handler that does something
- Suggestions loading history and hanging somehow there

Can you trigger it without any of your configuration? `env XDG_CONFIG_HOME=(mktemp -d) fish`?

**Comment by zx8** [comment]
I can't trigger it without any of my configuration.

That said, I have managed to reliably trigger some kind of weird behaviour by typing `git commit` (`git` isn't an alias) and hitting the spacebar a few times. Upon doing so, my terminal session descends into a madness that can only explain with a video...

* https://streamable.com/p1a71

I recorded it in a rush so the keystrokes are quite fast and hard to follow, but keep a close eye on the keystrokes I'm hitting in the bottom right. The weirdness starts to occur immediately after I type `git` and hit `<space>` a few times (`<space>` is shown as `[_]`). At the end of the video, I hit both `CTRL+C` and `CTRL+\` and nothing happens.

The only thing I can think of is my "custom" git completions:

```
$ cat ~/.config/fish/completions/git.fish
source /usr/local/share/fish/completions/git.fish

complete -f -c git -n '__fish_git_using_command delete-branch' -a '(__fish_git_branches)'
complete -f -c git -n '__fish_git_using_command delete-tag ' -a '(__fish_git_tags)'
```

I'll experiment a bit more and get back to you.

**Comment by faho** [comment]
>At the end of the video, I hit both CTRL+C and CTRL+\ and nothing happens.

Note that fish doesn't react to either by quitting or anything. Try ctrl-d, or escape or something, just to see if you're in the wrong mode.

Also, as a weird guess, try pasting something. It's possible you've triggered bracketed paste, in which case nothing but the bracketed paste end sequence would get you out it, and with the exception of control characters everything else would just be entered verbatim.

**Comment by floam** [comment]
>  in which case nothing but the bracketed paste end sequence would get you out it, and with the exception of control characters everything else would just be entered verbatim.

It seems like it would be a good idea to do something that could indicate to a person that this is the state the terminal is stuck in. Perhaps getting something like "Bracketing pasting..." into the terminal title until it is finished?

**Comment by floam** [comment]
Or: right as we get into bracketed paste mode, and until `__fish_disable_bracketed_paste`, we create a SIGINT handler that disables bracketed paste.

**Comment by zx8** [comment]
@faho `CTRL+D` and pasting seem to have to effect: https://streamable.com/9h1j4

The odd part is that when I pasted your message, the spaces were stripped out?

**Comment by faho** [comment]
>Perhaps getting something like "Bracketing pasting..." into the terminal title until it is finished?

Pasting usually is over in a matter of milliseconds. Getting *any* message to display would have the terminal flicker in some manner.

>Or better: right when we get into bracketed paste mode, and until __fish_disable_bracketed_paste, we create a SIGINT handler that disables bracketed paste.

Wouldn't someone then be able to include a \cc in text and bypass bracketed paste that way? Bracketed paste means you get the start sequence, then *random data* and then the end sequence. Well-behaved terminals remove the end sequence from the data, but nothing else is guaranteed.

Really... we've not had a single report that bracketed paste ever got stuck for anyone. It hasn't gotten stuck here either, or pasting something else would have moved out of it.

If you really want, how about adding a way to time out? If bracketed paste hasn't gotten any input after a second, end it?

But either way, that's a discussion for another issue, since paste apparently isn't at fault here.

>This is getting very weird. I just ran the following and can still reproduce the issue:

Okay, I'm assuming you've started a new fish after that, in an independent terminal (not as a child), and you can still reproduce it?

Let's try to get some more data. The output of the following would be nice:

- `set -g`
- `set -U`
- `functions --handlers`

**Comment by zx8** [comment]
```
zx8@mac ~> set -g
Apple_PubSub_Socket_Render /private/tmp/com.apple.launchd.TJCvbeC9dX/Render
CMD_DURATION 0
COLORFGBG '15;0'
COLORTERM truecolor
COLUMNS 172
FISH_VERSION 3.0.2-1146-g11209b75-dirty
HOME /Users/zx8
IFS \n\ \t
ITERM_PROFILE Default
ITERM_SESSION_ID w0t1p1:F2CD712A-DF5E-441D-859B-7A5A6786BB08
LANG en_GB.UTF-8
LC_TERMINAL iTerm2
LC_TERMINAL_VERSION 3.3.0beta8
LINES 49
LOGNAME zx8
PATH '/usr/bin'  '/bin'  '/usr/sbin'  '/sbin'  '/usr/local/bin'  …
PWD /Users/zx8
SHELL /usr/local/bin/fish
SHLVL 1
SSH_AUTH_SOCK /private/tmp/com.apple.launchd.bJsOaEh9Ch/Listeners
TERM xterm-256color
TERM_PROGRAM iTerm.app
TERM_PROGRAM_VERSION 3.3.0beta8
TERM_SESSION_ID w0t1p1:F2CD712A-DF5E-441D-859B-7A5A6786BB08
TMPDIR /var/folders/2r/fs7qk8ts6250vqgckkq90h8w0000gn/T/
USER zx8
XPC_FLAGS 0x0
XPC_SERVICE_NAME 0
_ set
__CF_USER_TEXT_ENCODING 0x1F5:0:0
__fish_active_key_bindings fish_default_key_bindings
__fish_added_user_paths
__fish_bin_dir /usr/local/Cellar/fish/HEAD-11209b7/bin
__fish_config_dir /Users/zx8/.config/fish
__fish_config_interactive_done
__fish_data_dir /usr/local/Cellar/fish/HEAD-11209b7/share/fish
__fish_help_dir /usr/local/Cellar/fish/HEAD-11209b7/share/doc/fish
__fish_last_bind_mode default
__fish_sysconf_dir /usr/local/Cellar/fish/HEAD-11209b7/etc/fish
__fish_user_data_dir /Users/zx8/.local/share/fish
fish_bind_mode default
fish_complete_path '/Users/zx8/.config/fish/completions'  '/usr/local/Cellar…
fish_function_path '/Users/zx8/.config/fish/functions'  '/usr/local/Cellar/f…
fish_pid 21586
fish_private_mode
fish_term24bit 1
history 'functions --handlers'  'mv ~/.local/share/fish{,.bak}'  'se…
hostname mac.local
pipestatus 0
status 0
umask 0022
version 3.0.2-1146-g11209b75-dirty
```

---

```
zx8@mac ~> set -U
__fish_init_2_39_8
__fish_init_2_3_0
__fish_init_3_x
fish_color_autosuggestion '555'  'brblack'
fish_color_cancel -r
fish_color_command 005fd7
fish_color_comment 990000
fish_color_cwd green
fish_color_cwd_root red
fish_color_end 009900
fish_color_error ff0000
fish_color_escape 00a6b2
fish_color_history_current --bold
fish_color_host normal
fish_color_match --background=brblue
fish_color_normal normal
fish_color_operator 00a6b2
fish_color_param 00afff
fish_color_quote 999900
fish_color_redirection 00afff
fish_color_search_match 'bryellow'  '--background=brblack'
fish_color_selection 'white'  '--bold'  '--background=brblack'
fish_color_user brgreen
fish_color_valid_path --underline
fish_greeting 'Welcome to fish, the friendly interactive shell'
fish_key_bindings fish_default_key_bindings
fish_pager_color_completion
fish_pager_color_description 'B3A06D'  'yellow'
fish_pager_color_prefix 'white'  '--bold'  '--underline'
fish_pager_color_progress 'brwhite'  '--background=cyan'
```

---

```
zx8@mac ~> functions --handlers
Event signal
SIGTRAP fish_sigtrap_handler
SIGWINCH __fish_winch_handler

Event variable
fish_color_cwd __fish_repaint
fish_color_cwd_root __fish_repaint
fish_key_bindings __fish_reload_key_bindings
fish_user_paths __fish_reconstruct_path

Event generic
fish_command_not_found __fish_command_not_found_handler
fish_exit __fish_disable_bracketed_paste
fish_preexec __fish_disable_bracketed_paste
fish_prompt __fish_enable_bracketed_paste
```

**Comment by faho** [comment]
Okay, that output is *entirely* normal. There's nothing even slightly out of the ordinary there.

But space not working brings me to a half-theory: It seems like it's not *executing* anything but the "generic" binding.

Because space is actually bound now, and no longer uses the generic "catch-all" binding of "self-insert". This is so it can trigger abbreviations.

@zx8: It might be possible to confirm with `bind --erase --preset " "`. That should still insert spaces, and should even do it when the weirdness happens. It won't expand abbreviations anymore, though.

**Comment by zx8** [comment]
Ok, so I launched a new session and executed `bind --erase --preset " "` straightaway, while I was still able to.

I then triggered the "bug" again by typing `git commit` and hitting backspace/space a few times (which seems to be the easiest way to trigger it).

After it was triggered, the session was still behaving strangely (as shown in the videos), but crucially, this time, hitting `<space>` _worked_ – so it would seem we're on to something now and it's somehow related to bindings...

**Comment by floam** [comment]
> Wouldn't someone then be able to include a \cc in text and bypass bracketed paste that way? 

I don't think so. Doesn't ^C actually send a signal, not text?

**Comment by faho** [comment]
>After it was triggered, the session was still behaving strangely (as shown in the videos), but crucially, this time, hitting <space> worked – so it would seem we're on to something now and it's somehow related to bindings...

Okay. So this means *something*, but I'm not yet sure what it is.

I assumed this was specific to the "generic" binding, but looking at the code again it seems unlikely. Because what happens is not that bindings are forgotten or overlooked, but matched and then not executed - fish gets that you entered a space, it just doesn't do the space thing. And the only special thing about the "generic" binding is that it's empty, so it matches everything.

So my theory now is that we don't execute anything that isn't a bind-function.

So what I'd expect to work are bindings like

```fish
bind \x7f backward-delete-char # backspace
bind \ef forward-word # alt+f
```

but what wouldn't work is something like

```fish
bind \x7f "commandline -f backward-delete-char"
bind \ef "commandline -f forward-word"
```

We only execute "commands" (i.e. not-bind-functions) in certain circumstances, so it's possible that gets stuck somehow.

**Comment by faho** [comment]
>I don't think so. Doesn't ^C actually send a signal, not text?

@floam: As a simple test, try pasting `echo \cc | pbcopy` back. For me, that triggers a warning message asking me if I really want to paste it, and if I do, it triggers the signal.

Which means it would be a way to get around bracketed paste's safety, like I suspected.

**Comment by ronjouch** [comment]
I confirm the bug on my work machine (Ubuntu 19.04), and I confirm rolling back to ff3fe96 fixes the issue. I'm following the discussion and will try reproducing the bug without my config.

**Comment by floam** [comment]
>  As a simple test, try pasting echo \cc | pbcopy back.

Ah, because it's not the terminal sending SIGINT, it's the line discipline.

**Comment by faho** [comment]
>And this is pretty off-topic, however on my terminal it just causes a new prompt to be printed with [201~ in the command line.

The bracketed paste end sequence is `\e\[201~`. If that shows up in the commandline, that quite literally means you've exited bracketed paste. So we actually *do* read the SIGINT, and we actually *do* exit paste mode.

AKA it's another bug.

**Comment by floam** [comment]
So actually `\cc` can be used to bypass the feature and execute stuff.

**Comment by ronjouch** [comment]
This looks fixed to me. I've been running fish 9ae3b345 for one work day and didn't encounter the problem (it used to appear after a few minutes of usage)

**Comment by faho** [comment]
@zx8: Can you confirm?

**Comment by zx8** [comment]
Interesting. I've been running https://github.com/fish-shell/fish-shell/commit/ff3fe961f4aa8d997ea2f65e309162161fc7df71 since I originally raised this issue, but I just updated to https://github.com/fish-shell/fish-shell/commit/7fe93c75186826a932954a9bf208f0aed5dc545b and can't reproduce the original reported bug no matter how hard I try, so it certainly seems to have been resolved in one of the more recent commits, but I'm not sure which one and I'm not sure it's worth the effort to try and figure it out.

Happy to resolve this and reopen if rears its head again in future!
